### PR TITLE
State refactor and logical and in rules.

### PR DESCRIPTION
**Pull request by kdrabek-form3**
Tests are failing because https://github.com/form3tech-oss/github-team-approver-commons/pull/11 has to be merged first.

This PR:
- introduces a logical "AND" between `regex`, `regex_label` and `directories` in a given pull_request_approval_rule.
  - a pull request must meet all conditions expressed in `regex`, `regex_label` and `directories` for it to match the rule. Empty fields are of course skipped
  - this can be a breaking change depending on the user's config
- adds support for detecting co-authors and optionally ignoring their approval
- changes implementation of tracking approvals for matched `pull_request_approval_rules`

### CODE REVIEW

File: internal/api/approval/approval.go [hunk 1]
```diff
@@ -400,3 +425,21 @@ func filterAllowedAndIgnoreReviewers(members []*github.User, commits []*github.R
 
 	return allowed, ignored
 }
+
+func findCoAuthors(msg string) []string {
+	pattern := "Co-authored-by: .+? <([\\w\\+-]+)@users.noreply.github.com>"
+	r := regexp.MustCompile(pattern)
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/approval.go:431]: nit, but it is pointless to compile pattern every time `findCoAuthors` is invoked

File: internal/api/approval/approval_test.go [hunk 2]
```diff
@@ -135,6 +136,57 @@ func TestContentsUrlToRelDir(t *testing.T) {
...
+			Co-authored-by: Jane Doe <jane-doe@users.noreply.github.com>`,
+			expected: []string{"john-doe", "jane-doe"},
+		},
+		"when one of the co-authors uses unsupported format": {
+			message: `feat: awesome new feature
+
+			Co-authored-by: John Doe <john@doe.com>
+			Co-authored-by: Jane Doe <87654321+jane-doe@users.noreply.github.com>`,
+			expected: []string{"jane-doe"},
+		},
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/approval_test.go:180]: I would add one more scenario:

```go
		"when duplicated co-author in commit message": {
			message: `feat: awesome new feature
			Co-authored-by: John Doe <12345678+john-doe@users.noreply.github.com>
			Co-authored-by: John Doe <12345678+john-doe@users.noreply.github.com>
			Co-authored-by: John Doe <12345678+john-doe@users.noreply.github.com>
			Co-authored-by: John Doe <12345678+john-doe@users.noreply.github.com>`,
			expected: []string{"???"},
		},
```
WDYT?

File: internal/api/approval/approval_test.go [hunk 3]
```diff
@@ -135,6 +136,57 @@ func TestContentsUrlToRelDir(t *testing.T) {
...
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			require.Equal(t, findCoAuthors(tt.message), tt.expected)
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/approval_test.go:185]: Imho safer to use [ElementsMatch](https://pkg.go.dev/github.com/stretchr/testify/require#ElementsMatch) as it ignores the order of elements.

File: internal/api/approval/approval_test.go [hunk 4]
```diff
@@ -191,15 +243,38 @@ func TestFilterAllowedAndIgnoreReviewers(t *testing.T) {
...
+		},
 	}
 
 	for name, tt := range tests {
 		t.Run(name,
 			func(t *testing.T) {
 				gotAllowed, gotIgnored := filterAllowedAndIgnoreReviewers(tt.members, tt.commits)
+				sort.Strings(gotAllowed)
+				sort.Strings(gotIgnored)
 				require.Equal(t, gotAllowed, tt.allowed)
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/approval_test.go:278]: if `ElementsMatch` assertion were used it is not necessary to sort slice

File: internal/api/approval/matched_rule.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	for _, count := range ta {
+		total += count
+	}
+	return total >= 1
+}
+
+func (ta TeamApprovals) AllTeamsApproved(allTeams []string) bool {
+	for _, handle := range allTeams {
+		approvals, present := ta[handle]
+		if !present || approvals < 1 {
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/matched_rule.go:80]: nit, but for the sake of readability I would prefer splitting conditions

```go
if !present {
   return false
}

if approvals < 1 {
   return false
}
```

File: internal/api/approval/matched_rule_test.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+			matchedRule: MatchedRule{
+				Approvals: TeamApprovals{},
+			},
+			result: []string{},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			require.Equal(t, tt.result, tt.matchedRule.ApprovingTeamNames())
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/matched_rule_test.go:108]: again, ElementsMatch would be more appropriate here

File: internal/api/approval/state.go [hunk 7]
```diff
@@ -87,41 +77,85 @@ func (s *state) addIgnoredReviewers(ignoredTeamMembers []string) {
...
+		if !rule.Fulfilled() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s *state) pendingTeamNames() []string {
+	allPending := []string{}
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/state.go:91]: isn't it more go-idiomatic to not instantiate until necessary :)?
```suggestion
	var allPending []string = nil
```

File: internal/api/approval/state.go [hunk 8]
```diff
@@ -87,41 +77,85 @@ func (s *state) addIgnoredReviewers(ignoredTeamMembers []string) {
...
 		result.status = StatusEventStatusError
-	case s.forceApproval:
+	case s.shouldForceApprove():
 		// The PR is being forcibly approved.
 		result.description = statusEventDescriptionForciblyApproved
 		result.status = StatusEventStatusSuccess
-		result.reviewsToRequest = computeReviewsToRequest(log, teams, s.pendingTeamNames)
-	case len(s.pendingTeamNames) > 0:
+		result.reviewsToRequest = computeReviewsToRequest(log, teams, pendingTeamNames)
+	case len(pendingTeamNames) > 0 && !allRulesFulfilled:
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/state.go:145]: why new AND condition is added here? 🤔

File: internal/api/approval/state.go [hunk 9]
```diff
@@ -87,41 +77,85 @@ func (s *state) addIgnoredReviewers(ignoredTeamMembers []string) {
...
 		result.status = StatusEventStatusPending
-		result.reviewsToRequest = computeReviewsToRequest(log, teams, s.pendingTeamNames)
-	case len(s.pendingTeamNames) == 0 && len(s.approvingTeamNames) == 0:
+		result.reviewsToRequest = computeReviewsToRequest(log, teams, pendingTeamNames)
+	case len(pendingTeamNames) == 0 && len(approvingTeamNames) == 0:
 		// No teams have been identified as having to be requested for a review.
 		// NOTE: This should not really happen in practice.
 		result.description = statusEventDescriptionNoReviewsRequested
 		result.status = StatusEventStatusSuccess
-	default:
```
Comment (wojciecholszewski-form3) [inline internal/api/approval/state.go:122]: what could be the implications of removing `default`?

File: internal/api/pacts/pull_request_review_submitted_approval_mode_require_any.json [hunk 10]
```diff
@@ -79,7 +79,7 @@
         "headers": {
           "Content-Type": "text/plain; charset=utf-8"
         },
-        "body": "pull_request_approval_rules:\n- target_branches:\n  - master\n  rules:\n  - approval_mode: require_any\n    regex: \"- \\\\[x\\\\] Yes - this change impacts customers\"\n    approving_team_handles:\n    -  CAB - Foo\n    -  CAB - Bar\n    labels:\n    - needs-cab-approval\n  - regex: \"- \\\\[x\\\\] Yes - this change impacts documentation\"\n    approving_team_handles:\n    -  CAB - Documentation\n    labels:\n    - needs-doc-approval\n"
+        "body": "pull_request_approval_rules:\n- target_branches:\n  - master\n  rules:\n  - approval_mode: require_any\n    regex: \"- \\\\[x\\\\] Yes - this change impacts customers\"\n    approving_team_handles:\n    -  CAB - Foo\n    -  CAB - Bar\n    labels:\n    - needs-cab-approval\n  - approval_mode: require_all\n    regex: \"- \\\\[x\\\\] Yes - this change impacts documentation\"\n    approving_team_handles:\n    -  CAB - Documentation\n    labels:\n    - needs-doc-approval\n"
```
Comment (wojciecholszewski-form3) [inline internal/api/pacts/pull_request_review_submitted_approval_mode_require_any.json:82]: I am confused by setting `approval_mode: require_all`. Pact name `internal/api/pacts/pull_request_review_submitted_approval_mode_require_any.json` indicates mode should be `require_any`, no?

File: internal/api/stages/fakegithub/github.go [hunk 11]
```diff
@@ -111,7 +119,7 @@ func (f *FakeGitHub) SetCommits(r []*github.RepositoryCommit) {
 }
 
 func (f *FakeGitHub) SetReviews(r []*github.PullRequestReview) {
-	f.reviews = r
+	f.reviews = append(f.reviews, r...)
```
Comment (wojciecholszewski-form3) [inline internal/api/stages/fakegithub/github.go:122]: why appending instead of overwriting?

**Review by wojciecholszewski-form3 — COMMENTED** [review]
Few comments left. I am not approving yet.

File: internal/api/approval/matched_rule.go (hunk 5 above)
Comment (kdrabek-form3) [reply internal/api/approval/matched_rule.go:80]: To me if statement with one simple `or` condition still seems readable, but I don't really have a preference. Will change.

File: internal/api/approval/approval_test.go (hunk 2 above)
Comment (kdrabek-form3) [reply internal/api/approval/approval_test.go:180]: Added. Had to slightly change the function implementation as it didn't handle duplicated very well

Comment (kdrabek-form3) [reply internal/api/approval/approval_test.go:180]: Added. Had to slightly change the function implementation as it didn't handle duplicated very well

File: internal/api/stages/fakegithub/github.go (hunk 11 above)
Comment (kdrabek-form3) [reply internal/api/stages/fakegithub/github.go:122]: Replacing prevents us from chaining approval stage functions such as `AliceApprovesPullRequest` and `BobApprovesPullRequest`

File: internal/api/pacts/pull_request_review_submitted_approval_mode_require_any.json (hunk 10 above)
Comment (kdrabek-form3) [reply internal/api/pacts/pull_request_review_submitted_approval_mode_require_any.json:82]: It should. It was a `require_all` with one approving team handle anyway. Copy-paste mistake.

File: internal/api/approval/state.go (hunk 9 above)
Comment (kdrabek-form3) [reply internal/api/approval/state.go:122]: Assuming no other cases match, I think we will ultimately log `failed to handle event` and the pull request status will never change. 

But I was a bit concerned about defaulting to `success` to be honest. I think I'd rather log error than assume the pull request is good to go.

File: internal/api/approval/state.go (hunk 7 above)
Comment (kdrabek-form3) [reply internal/api/approval/state.go:91]: I'm sure you're right but I need empty list to return as default in case there are no pending team names.

**Review by wojciecholszewski-form3 — APPROVED** [review]
Thanks for taking my remarks into account! LGTM 👍

**Pull request merged**